src/components/QueryBuilder/components/JsonLogicEditor.tsx

```diff
@@ -0,0 +1,81 @@
+import React, { FunctionComponent, useState } from 'react';
+import { Button, Box } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+import { JsonLogicTree } from 'react-awesome-query-builder';
+import classNames from 'classnames';
+
+import { useSafeIntl } from '../../../utils/useSafeIntl';
+import { FormControl } from '../../inputs/FormControl';
+import { MESSAGES } from '../messages';
+
+const styles = theme => ({
+    textArea: {
+        width: '100%',
+        minWidth: '100%',
+        minHeight: '100px',
+        padding: theme.spacing(2),
+        outline: 'none',
+    },
+    textAreaError: {
+        borderColor: `${theme.palette.error.main} !important`,
+    },
+});
+
+// @ts-ignore
+const useStyles = makeStyles(styles);
+
+type Props = {
+    initialLogic?: JsonLogicTree;
+    // eslint-disable-next-line no-unused-vars
+    changeLogic: (newLogic: JsonLogicTree) => void;
+};
+
+export const JsonLogicEditor: FunctionComponent<Props> = ({
+    initialLogic,
+    changeLogic,
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const classes: Record<string, string> = useStyles();
+    const [logic, setLogic] = useState<string>(JSON.stringify(initialLogic));
+    const [errors, setErrors] = useState<Array<string>>([]);
+
+    const handleChange = e => {
+        const { value } = e.target;
+        setLogic(value);
+        try {
+            JSON.parse(value);
+            setErrors([]);
+        } catch {
+            setErrors([formatMessage(MESSAGES.invalidJson)]);
+        }
+    };
+
+    return (
+        <Box mt={2} mb={-2}>
+            <FormControl errors={errors}>
+                <textarea
+                    className={classNames(
+                        classes.textArea,
+                        errors.length > 0 && classes.textAreaError,
+                    )}
+                    onChange={handleChange}
```

Comment:
have you tested switching languages. It seems to me that since `handleChange` is not wrapped in a `useCallback` you should have problems with the translation of errors when you switch languages, because of the dep to `formatMessage`